proposals/2020-06-26_decrypt-max-header-size-max-body-size/proposal.md

```diff
@@ -0,0 +1,275 @@
+    represents these objects in memory.
+    These objects MUST be held in memory for the duration of the decryption operation
+    because the decrypt operation MUST output the parsed header.
+
+Messages with the above fields at their size limits, when processed,
+can exhaust many memory runtime limits.
+As such, customers who are concerned with the stability and memory usage of their systems
+may care about providing stricter bounds to these fields.
+
+The Encryption Context field in the header
+and the Signature Length field in the footer are also variable length
+but they cannot exceed 2^16 - 1 bytes (64 KiB).
+This is well within the bounds of memory runtime limits for the most languages,
+so we do not consider the bounds of these fields a concern.
+
+Note that all implementations dynamically allocate resources while processing messages.
+
+### When should users set this value?
+
+Users should set this value if they need to process
+messages of unknown size
+or messages from an untrusted source.
+
+### When should users NOT set this value?
+
+This control should be considered strictly an operational control that can
+mitigate highly-bounded resource consumption for certain use cases.
+
+If a user has a security requirement to not decrypt messages with a certain content length
+or message header length
+(ex, to enforce a “sign only” use case by only decrypting zero length messages),
+that security requirement should be met
+with a custom CMM that only allows desired requests
+rather than using these controls.
+
+### What value should they set it to?
+
+Users should choose reasonable values for their use case and
+the messages that they expect to decrypt.
+They should expect the memory usage of the ESDK to scale linearly with this control,
+and can performance test their application to determine
+the ESDKs memory cost in practice for their use case.
+
+### Why not just provide one control that “does the right thing” for most use cases?
+
+The two controls mitigate two different resource consumption concerns.
+`Max Body Size` limits short term memory consumption.
+This can be used to ensure that the decryption of a frame or
+an unframed message doesn't immediately exhaust memory linear to a factor of that value.
+
+`Max Header Size` has a fuzzier implication on memory that
+depends on the implementation
+and represents a memory requirement that exists for the duration of the decryption operation.
+
+Separating these controls
+keeps the purpose and affect of each control clear and simple,
+so that they can be easily understood and applied.
+
+### Should there be a lower limit for Max Header Size?
+
+To allow for the smallest possible message header
+(no AAD, one EDK but the EDK fields themselves are empty),
+the `Max Header Size` MUST NOT be less than 40 bytes.
+Because future message format revisions might have different lower limits,
+we cannot know if a limit is too low until we start processing a message.
```

Comment:
```suggestion
However, because future message format revisions might have different lower limits,
we cannot know if a limit is too low until we start processing a message.
```